### PR TITLE
JAMES-1785 setInMailboxes move a message

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java [hunk 1]
```diff
@@ -19,15 +19,141 @@
...
 
 public class SystemMailboxesProviderImplTest {
 
-    @Ignore("1716 this class needs a test suite")
+    private static final MailboxPath INBOX = MailboxManagerFixture.MAILBOX_PATH1;
+    private static final MailboxPath OUTBOX = MailboxManagerFixture.MAILBOX_PATH2;
+    private static final char DELIMITER = '.';
+
+    private static MailboxId inboxId = TestId.of(1);
+    private static MailboxId outboxId = TestId.of(1);
```
Comment (chibenwa) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java]: Two times the same ID :trollface: ?

Comment (chibenwa) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java]: Should be final

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java [hunk 2]
```diff
@@ -19,15 +19,141 @@
...
+
+    @Mock private MessageManager inboxMessageManager;
+    @Mock private MessageManager outboxMessageManager;
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    @Before
+    public void setUp() throws Exception {
+        MockitoAnnotations.initMocks(this);
```
Comment (chibenwa) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java]:
```
@RunWith(MockitoJUnitRunner.class)
```

On top of the class...

http://stackoverflow.com/questions/15494926/initialising-mock-objects-mockito

This will make this call avoidable :-)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java [hunk 3]
```diff
@@ -54,13 +61,21 @@
...
+    private final SystemMailboxesProvider systemMailboxesProvider;
 
     @Inject
     @VisibleForTesting SetMessagesUpdateProcessor(
-            UpdateMessagePatchConverter updatePatchConverter,
-            MessageIdManager messageIdManager) {
+        UpdateMessagePatchConverter updatePatchConverter,
+        MessageIdManager messageIdManager,
+        Factory mailboxIdFactory,
+        MailboxManager mailboxManager, SystemMailboxesProvider systemMailboxesProvider) {
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: Break line before SystemMailboxesProvider

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 4]
```diff
@@ -39,7 +39,9 @@
 import org.apache.james.mailbox.MailboxSession;
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: WIP GetMessages looks like a terrible commit message... I wonder who did it... Could you fix it ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 5]
```diff
@@ -1304,43 +1305,43 @@ public void setMessagesShouldKeepBccInSentMailbox() throws Exception {
...
-                "        \"subject\": \"Thank you for joining example.com!\"," +
-                "        \"textBody\": \"Hello someone, and thank you for joining example.com!\"," +
-                "        \"mailboxIds\": [\"" + getOutboxId(accessToken) + "\"]" +
-                "      }}" +
-                "    }," +
-                "    \"#0\"" +
-                "  ]" +
-                "]";
+            "  [" +
+            "    \"setMessages\","+
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: A hudge reindent should be in a different commit than code addition...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 6]
```diff
@@ -99,16 +100,17 @@ public Builder threadId(String threadId) {
...
 
+        @JsonIgnore
         public Builder mailboxId(MailboxId mailboxId) {
-            return this.mailboxIds(ImmutableList.of(mailboxId));
+            return this.fluentMailboxIds(mailboxId);
         }
 
         @JsonIgnore
-        public Builder mailboxIds(MailboxId... mailboxIds) {
+        public Builder fluentMailboxIds(MailboxId... mailboxIds) {
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: Just to say it, a different method name is required, otherwise Jackson fails. Hence, no mailboxIds...

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 5 above)
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: yes, and some reindent are wrong (.when() / .then())
please fix this so we can review it finely

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 7]
```diff
@@ -1677,6 +1679,45 @@ public void movingShouldBeSupported() throws Exception {
                         + "(through reference chain: org.apache.james.jmap.model.Builder[\"mailboxIds\"])"))
                .body(ARGUMENTS + ".updated", hasSize(0));
     }
+
+    @Test
+    public void moveToTrashIsNotYetSupported() throws Exception {
+        String newMailboxName = "heartFolder";
+        jmapServer.serverProbe().createMailbox("#private", USERNAME, newMailboxName);
```
Comment (strokyl) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: I don't understand why you create such a mailbox and then never use it in your test

**Comment by chibenwa** [comment]
```
[Building Java 8 james-1785-21147b06-6797-4084-88d1-6d41fc53818b] Tests run: 132, Failures: 0, Errors: 2, Skipped: 2, Time elapsed: 63.041 sec <<< FAILURE! - in org.apache.james.jmap.cassandra.cucumber.CassandraDownloadCucumberTest
[Building Java 8 james-1785-21147b06-6797-4084-88d1-6d41fc53818b] Given "usera@domain.tld" is trusted for attachment "a1"(Scenario: A user should not have access to the download endpoint without a blobId)  Time elapsed: 0.011 sec  <<< ERROR!
```

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProvider.java [hunk 8]
```diff
@@ -28,4 +28,6 @@
 
 public interface SystemMailboxesProvider {
     Stream<MessageManager> listMailboxes(Role aRole, MailboxSession session) throws MailboxException;
+
+    MessageManager findMailbox(Role role, MailboxSession session) throws MailboxException;
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProvider.java]: We tends to avoid default implementation in interfaces, as far as I remember, but yes, you are right.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 7 above)
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Copy and paste mistake ? :-p

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java (hunk 1 above)
Comment (quynhn) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java]: Done

**Comment by chibenwa** [comment]
We need to modify JMAP advertised capabilities : 

james-project/server/protocols/jmap/doc/specs/spec/message.mdwn

line 250 remove "not implemented"

However we need a "not implemented" comment line 330:

```
To **delete a message** to trash, simply change the `mailboxIds` property so it is now in the mailbox with `role == "trash"`.
```

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java (hunk 2 above)
Comment (quynhn) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java]: Yes, we can but
http://stackoverflow.com/questions/10806345/runwithmockitojunitrunner-class-vs-mockitoannotations-initmocksthis
It said:
```
Using runner lets you save a little bit of coding (no need for @Before method). On the other hand using a runner is sometimes not possible, i.e. when you are already using one, like SpringJUnit4ClassRunner.
```
Then I think we keep as it is

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java (hunk 3 above)
Comment (quynhn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: Yes

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java (hunk 4 above)
Comment (quynhn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: Yes, then I changed

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 5 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: I separated them all

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProvider.java (hunk 8 above)
Comment (quynhn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProvider.java]: Done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 7 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: And removed

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 9]
```diff
@@ -1706,20 +1743,254 @@ public void moveToTrashIsNotYetSupported() throws Exception {
...
             .body(requestBody)
-        .when()
+            .when()
+            .post("/jmap");
+
+        String firstMessage = ARGUMENTS + ".list[0]";
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body("[[\"getMessages\", {\"ids\": [\"" + messageToMoveId + "\"]}, \"#0\"]]")
+            .when()
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Indent on When

Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: And elsewhere in this commit

Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: What if it had a separated commit for fixing indent issues?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/send/PostDequeueDecorator.java [hunk 10]
```diff
@@ -95,7 +96,7 @@ private boolean checkMessageIdAttribute() {
         try {
             messageIdFactory.fromString((String) messageId);
         } catch (Exception e) {
-            LOG.error("Invalid messageId: " + (String) messageId);
+            LOG.error("Invalid messageId: " + messageId);
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/send/PostDequeueDecorator.java]: why this change?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java [hunk 11]
```diff
@@ -41,7 +42,7 @@
     private final MailboxManager mailboxManager;
 
     @Inject
-    @VisibleForTesting SystemMailboxesProviderImpl(MailboxManager mailboxManager) {
+    public SystemMailboxesProviderImpl(MailboxManager mailboxManager) {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java]: why this change?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessorTest.java [hunk 12]
```diff
@@ -66,7 +66,7 @@ public void processShouldReturnNonEmptyNotUpdatedWhenRequestHasInvalidUpdate() {
         when(mockConverter.fromJsonNode(any(ObjectNode.class)))
                 .thenReturn(mockInvalidPatch);
 
-        SetMessagesUpdateProcessor sut = new SetMessagesUpdateProcessor(mockConverter, null);
+        SetMessagesUpdateProcessor sut = new SetMessagesUpdateProcessor(mockConverter, null, null, null, null);
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessorTest.java]: Is this related to this commit?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java (hunk 2 above)
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java]: usually we don't use the @Mock annotation but more the mock() static invocation. Is there any reason to change?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 13]
```diff
@@ -1641,62 +1642,98 @@ private boolean isTextPlusHtmlMessageReceived(AccessToken recipientToken) {
...
+            .body(requestBody)
+            .when()
+            .post("/jmap")
+            .then()
+            .log().ifValidationFails()
+            .spec(getSetMessagesUpdateOKResponseAssertions(messageToMoveId));
+    }
+
+    @Test
+    public void mailboxIdsShouldBeInDestinationWhenUsingForMove() throws Exception {
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: maybe you are missing a test: mailboxIdsShouldNotBeAnymoreInSourceWhenUsingForMove

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 14]
```diff
@@ -117,18 +119,28 @@ private GetMessagesResponse getMessagesResponse(MailboxSession mailboxSession, G
...
+                List<MailboxId> mailboxIds = messageResults.stream()
+                    .map(MessageResult::getMailboxId)
+                    .collect(Guavate.toImmutableList());
+                return messageResults.stream()
+                    .findFirst()
+                    .map(Stream::of)
+                    .orElse(Stream.of())
                     .map(Throwing.function(MetaDataWithContent::builderFromMessageResult).sneakyThrow())
                     .map(builder -> builder.messageId(messageId))
+                    .map(builder -> builder.mailboxIds(mailboxIds))
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: this logic should be in messageIdManager, no?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java [hunk 15]
```diff
@@ -55,8 +55,8 @@
     private static final MailboxPath OUTBOX = MailboxManagerFixture.MAILBOX_PATH2;
     private static final char DELIMITER = '.';
 
-    private static MailboxId inboxId = TestId.of(1);
-    private static MailboxId outboxId = TestId.of(1);
+    private static final MailboxId inboxId = TestId.of(1);
+    private static final MailboxId outboxId = TestId.of(2);
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java]: next constants can be final too

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/send/PostDequeueDecorator.java (hunk 10 above)
Comment (quynhn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/send/PostDequeueDecorator.java]: Because of warning about cast, we don't need to cast it

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java (hunk 11 above)
Comment (quynhn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java]: As I thought, we add "VisibleForTesting" if we only want to use/test that method. In this case, we do not testing constructor function so I removed it.
Do I need to roll it back?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessorTest.java (hunk 12 above)
Comment (quynhn) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessorTest.java]: Yes, it is. We use SetMessagesUpdateProcessor with one more SystemMailboxesProvider parameter constructor

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java (hunk 15 above)
Comment (quynhn) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/SystemMailboxesProviderImplTest.java]: Yes, I changed it

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java (hunk 11 above)
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java]: no, but then you cat make it private

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 13 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Yes and it's added

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessorTest.java (hunk 12 above)
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessorTest.java]: yes but here you have 3 more parameters, which seems to be more related to the next commit

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java (hunk 11 above)
Comment (quynhn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java]: I tried but later on, the constructor is used at PostDequeueDecoratorTest so it cannot be private. 
On PostDequeueDecoratorTest, I tried to make a mock for SystemMailboxesProviderImpl but it will be complicated after that.
How can I do in this case?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessorTest.java (hunk 12 above)
Comment (quynhn) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessorTest.java]: Got it, I've checked the history and now I know. I move this change to right commit

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java (hunk 11 above)
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/SystemMailboxesProviderImpl.java]: So if it is used for testing, just keep the @VisibleForTesting annotation :)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java (hunk 14 above)
Comment (quynhn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: Yes, I will try

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 9 above)
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: You still have indent issue : 

```
given()
    .header(...)
.when()
    .post()
.then()
    .statusCode(200)
```

And not : 

```

given()
    .header(...)
    .when()
    .post()
    .then()
    .statusCode(200)
```

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 16]
```diff
@@ -1639,71 +1643,467 @@ private boolean isTextPlusHtmlMessageReceived(AccessToken recipientToken) {
...
+            "      }}" +
+            "    }," +
+            "    \"#0\"" +
+            "  ]" +
+            "]";
+
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body(requestBody)
+            .when()
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Here too

Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Done

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please
